### PR TITLE
Java: smart type conversion

### PR DESCRIPTION
**Pull request by QianJin2013**
This is almost same as https://github.com/rethinkdb/rethinkdb/pull/5484, but resolved conflict.

Hope this easy to review and merge.

There are still something TODO as following, i will modify in this branch later
    Better conversion(db -> java):
       Support generic hash map, generic array, generic primitive array
       Better conversion for number type xxx by ((Number)value).xxxValue(), instead of xxx.parse(value.toString())
    Support save primitive array to db.
...

### CODE REVIEW

**Comment by deontologician** [comment]
Since #5618 was merged, now there's a conflict. It solved the missing timeout exception on the cursor

File: drivers/java/src/main/java/com/rethinkdb/annotations/IgnoreNullFields.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package com.rethinkdb.annotations;
+
+import java.lang.annotation.*;
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+/**
+ * When save this data to rethink db, skip null properties.
```
Comment (deontologician) [inline drivers/java/src/main/java/com/rethinkdb/annotations/IgnoreNullFields.java]: Small grammar fix: "When saving this data"

**Comment by QianJin2013** [comment]
OK, i'v solved conflict and fixed some grammar and push forcibly.

**Comment by QianJin2013** [comment]
I don't know why some unrelated C++ files is listed in changed files list.

**Comment by pires** [comment]
Because you seem to have merged instead of rebase `next` branch.

**Comment by QianJin2013** [comment]
@pires thank you.  I just checkout branch next from github.com/rethinkdb/rethinkdb, then modify it, then push it forcibly to branch next_smart_type_conversion of github.com/sjitech/rethinkdb. It seems i should rebase my branch to some revision then merge my modification again. I'd thought these two ways are same. Anyway, hope this request works.

**Comment by QianJin2013** [comment]
In my github, i only see 8 changed files instead of 12 files here.

https://github.com/rethinkdb/rethinkdb/compare/next...sjitech:next_smart_type_conversion?expand=1

**Comment by NeilHanlon** [comment]
@QianJin2013 You *might* want to close this PR and re-open one so that it's clean--especially since some of the core was changed in this PR (I would guess due to your merge & force push vs rebase).

**Comment by QianJin2013** [comment]
@NeilHanlon thanks. I also did like that but still same request get used.

Anyway if can not be merged by that reason i will create a new branch to do it again.

**Comment by deontologician** [comment]
Yeah, it looks like the rebase moved some rewrote some commits from next

**Comment by QianJin2013** [comment]
@deontologician so should i create another clean request?

**Comment by deontologician** [comment]
There are some windows files that get accidentally committed due to a
.gitattributes problem. I dont think you should open a new PR. Whatever you
could do by creating a new branch could be done by fixing the current one
and force pushing. In the worst case, you could set up another branch the
way you like it, then reset this branch to that one and force push it

**Pull request closed without merging**